oat-sa/extension-tao-itemqti-pci
# Fix/tr 848/allow partial audio recording

Related to: https://oat-sa.atlassian.net/browse/TR-848

Fix some runtime issues occurring when the user leaves the interaction without stopping either the recording or the playback.

Also, activate the option to allow partial updates of the response while the audio is recording. This feature may have a performance impact, hence it was hidden previously. To mitigate this performance impact, I added a throttling to the partial update, preventing it to happen too often. For each partial update, the whole record is supplied. Therefore, having a throttle should not be a problem.

![image](https://user-images.githubusercontent.com/1500098/114664634-44821700-9cfc-11eb-8a32-2d277f496d1f.png)

**How to test**
- **Prerequisites:**
    - make sure the PCI is updated, by either doing a fresh install or performing an update of the TAO instance
    - have several items with the audioRecording PCI. Ideally, make sure to have at least compressed and uncompressed versions.
    - have a test delivery with these items
    - the platform needs to be installed in secure mode (HTTPS)
- **Steps**
    - as a test taker, pass the test, and do the following (several test executions will be needed to check all):
        - for each item record some sound, play it back and move to the next item without stopping the playback: no error should be triggered in the console
        - for each item record some sound and move to the next item without stopping the recording: no error should be triggered in the console
    - as a back office user, go to the results page
        - for the test execution where the recordings have been completed before moving to the next item, all recording should be available (download and review)
        - for the test execution where the recordings have been truncated because a move was made without stopping them, the recordings should be available as well. However, the sound may not be complete as a few seconds of record may be missing at the end, depending on the configured update interval.

**Note:** the unit tests should pass. However, the change cannot be fully unit tested due to the nature of the feature which cannot be mocked or processed without user interaction. The unit test needs to be taken from an actual browser (the CLI mode is disabled mostly).
To launch the unit test, open your browser at the URL: `http://YOUR_HOST/qtiItemPci/views/js/test/audioRecordingInteraction/test.html`
